rpki-validator/src/main/java/net/ripe/rpki/validator3/domain/validation/RpkiRepositoryValidationService.java

```diff
@@ -333,26 +339,24 @@ private RpkiRepository findDownloadedParentRepository(Map<URI, RpkiRepository> f
     private void storeObjects(File targetDirectory,
                               RsyncRepositoryValidationRun validationRun,
                               ValidationResult validationResult,
-                              Set<String> existingObjectsKeys,
+                              Set<String> storedObjectKeys,
                               RpkiRepository repository) {
-        storage.writeTx0(tx -> {
-            try {
-                traverseFSandStore(tx, targetDirectory, validationRun, validationResult, existingObjectsKeys, repository);
-            } catch (IOException e) {
-                throw new RuntimeException(e);
-            }
-        });
+        try {
+            traverseFSandStore(targetDirectory, validationRun, validationResult, storedObjectKeys, repository, NO_MANIFESTS_PREDICATE);
+            traverseFSandStore(targetDirectory, validationRun, validationResult, storedObjectKeys, repository, ONLY_MANIFESTS_PREDICATE);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
     }
 
-    private void traverseFSandStore(Tx.Write tx,
-                                    File targetDirectory,
+    private void traverseFSandStore(File targetDirectory,
                                     RsyncRepositoryValidationRun validationRun,
                                     ValidationResult validationResult,
-                                    Set<String> existingObjectsKeys,
-                                    RpkiRepository repository) throws IOException {
-
-        final AtomicInteger workCounter = new AtomicInteger(0);
-        final int threshold = 10;
+                                    Set<String> storedObjectKeys,
+                                    RpkiRepository repository,
+                                    Predicate<RepositoryObjectType> typePredicate) throws IOException {
+        AtomicInteger pendingObjectsBytes = new AtomicInteger(0);
+        List<Pair<String, byte[]>> pendingObjects = new ArrayList<>(1000);
```

Comment:
Would it make sense to put filenames here instead of the objects and read them later?

Reply:
Also, since `asyncCreateObjects` is gone, how slower it is now?
I seems we save a lot of heap, but make it slower, how significant is the speed difference?

Reply:
Some timings from the log output. These timings are unreliable, so should only be used as a rough indication:

Current code on master:

```
Stored 11654 objects [...] in 2723ms
Stored 8381 objects [...] in 1503ms
```

Two FS traversal phases, no executor pool:

```
Stored 11651 objects [...] in 6797ms
Stored 8381 objects [...] in 4023ms
```

Singe FS traversal was not measurably different.

Two FS traversal phases with parallel stream for RPKI object conversion:

```
Stored 11651 objects [...] in 2964ms
Stored 8381 objects [...] in 1653ms
```

Above but first check for existence using SHA-256 hash without parsing the object at all:

```
Stored 11651 objects [...] in 2168ms
Stored 8381 objects [...] in 1088ms
```

For now I'll commit the parallel stream version as that is probably fast enough without adding additional code for existence checking.

Reply:
Cool!